index.bs

```diff
@@ -558,6 +563,52 @@ spec: SHA2; urlPrefix: http://csrc.nist.gov/publications/fips/fips180-4/fips-180
 
   <!-- ####################################################################### -->
 
+  # Authoring Considerations # {#authoring-considerations}
+
+  <em>This section is not normative.</em>
+
+  ## Versioned resources ## {#versioned-resources}
+
+  <a>Integrity metadata</a> is only useful if the resource can be reasonably
+  guaranteed to remain the same. It is therefore important that if resources
+  are updated they receive a new URL. Typically this might take the form of a
+  version number as part of the path or file name, which would ensure that a
+  resource would only be expected to change as a result of a compromised
+  server.
+
+  <div class="example">
+  <pre>
+    &lt;script src="https://example.com/v1.0/example-framework.js"
+            integrity="sha384-Li9vy3DqF8tnTXuiaAJuML3ky+er10rcgNR/VqsVpcw+ThHmYcwiB1pbOxEbzJr7"
+            crossorigin="anonymous"&gt;&lt;/script&gt;
+  </pre>
+  </div>
+
+  ## Coping with errors ## {#coping-with-errors}
+
+  When an integrity violation is detected, the resource will not parse or
+  execute. This could cause problems for users, so the system should be
+  designed to cope with errors.
+
+  If an alerting or logging system exists, the integrity error should be
+  sent to that so that the violation can be resolved safely.
+
+  The integrity error could also trigger the system to attempt to load the
+  resource from a local copy instead of a CDN or other third-party server.
+  It is worth evaluating if a local resource may be a better initial choice
+  rather than using a resource hosted on a CDN.
```

Comment:
Perhaps you could point to an example of this fallback mechanism? The note in https://www.w3.org/TR/SRI/#handling-integrity-violations tries to give a little more detail, but perhaps we could sketch something here that handles the `error` event?

Reply:
Honestly, I’d assumed this would be easier than it’s turned out to be. It’s easy to demonstrate this cleanly with `fetch`. Sample code:
```
let integrity = 'sha384-Li9vy3DqF8tnTXuiaAJuML3ky+er10rcgNR/VqsVpcw+ThHmYcwiB1pbOxEbzJr7';
fetch('https://example.com/v1.0/example-framework.js', { integrity })
	.then(handleResponse)
	.catch(error => {
		fetch('/jquery-3.6.0.slim.min.js')
			.then(handleResponse);
	});
```
But who’s going to be loading JS like that? Unfortunately the naïve approach of selecting an element and adding an `error` handler fails as the error fires too quickly (or there’s some other complication I’m not seeing).

Looking at how existing libraries to handle failing integrity work, they’ve doing fairly complex things with `MutationObserver`s (see https://sri.js.org/). Or at least too complex for a simple demo in a spec.

In the end, I wrote a event handler on `document` that captures all `error` events, and retargets them to a relative URL if they fail. I’m not used to writing spec-quality examplecode, so all criticism welcome. Does this look OK?
```
<script>
document.addEventListener('error', error => {
	const erroringScript = error.target;
	if (erroringScript.getAttribute('integrity')) {
		let fallbackURL = erroringScript.getAttribute('src').replace(/https:\/\/.*?\//, '/');
		let fallbackScript = document.createElement('script');
		fallbackScript.setAttribute('src', fallbackURL);
		erroringScript.after(fallbackScript);
	}
}, true);
</script>

<script src="https://example.com/v1.0/example-framework.js"
	integrity="sha384-Li9vy3DqF8tnTXuiaAJuML3ky+er10rcgNR/VqsVpcw+ThHmYcwiB1pbOxEbzJr7"
	crossorigin="anonymous"></script>
```

Reply:
I don't think we need a complete implementation.
It's worth pointing out that to keep the site functional, mismatching resources must be reloaded from trustworthy alternatives (e.g., a same-origin mirror) and that script elements need to be listened on early enough.

Reply:
`same-origin mirror` was the phrase my brain wasn’t providing, thanks.

I’ve pushed up something I think is reasonable now. The code example I’ve left as is as it’s at least working and somewhat compact. But if you think it’s too big for the spec then let’s just remove it and add a little more info to the following paragraph.